providers/implementations/ciphers/ciphercommon.c

```diff
@@ -435,19 +438,23 @@ int ossl_cipher_generic_block_final(void *vctx, unsigned char *out,
         return 0;
     }
 
-    if (ctx->pad && !ossl_cipher_unpadblock(ctx->buf, &ctx->bufsz, blksz)) {
-        /* ERR_raise already called */
-        return 0;
+    /* Note that ERR_raise is not called in case of bad padding */
+    pad = ~constant_time_is_zero(ctx->pad);
+
+    if (pad) {
+        gp = ~constant_time_is_zero(ossl_cipher_unpadblock(ctx->buf, &ctx->bufsz, blksz));
     }
 
+    ret = constant_time_select(pad, gp, CONSTTIME_TRUE);
+
     if (outsize < ctx->bufsz) {
         ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
         return 0;
     }
     memcpy(out, ctx->buf, ctx->bufsz);
```

Comment:
memcpy is not constant time either?

Reply:
Are you thinking about memcmp()?

Reply:
Yes, the length depends on the incorrect padding.
And is it possible that the if condition above (outsize < ctx->bufsz)
is satisfied in case of an incorrect padding?

Reply:
So bufsz is the size of the unpadded buffer, and this function should probably always work in blocksize. Which means the if above should also use the blocksize.

I assume this function can only deal with either 0 or 1 block.

Reply:
After unpadding (`ossl_cipher_unpadblock`), `ctx->bufsz` is always ≤ `blksz`. Regardless of whether the block is correctly padded, `outsize` is always ≥ `ctx->bufsz` since `outsize` is set in `EVP_DecryptFinal_ex` to either 0 or `blocksize` (`blksz`). So the if condition above (`outsize` < `ctx->bufsz`) is not satisfiable.

Reply:
About `memcpy`, I agree it is not constant time. It should probably be refactored as a constant-time memcpy that works on `blocksize` and masks the padding out from the output buffer, in a similar fashion to the constant-time loop in `ossl_cipher_unpadblock`.